File: tests/algorithms/x/PHP/data_structures/kd_tree/build_kdtree.php
<?php
ini_set('memory_limit', '-1');
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$tree = [];
function sort_points($points, $axis) {
  global $tree, $pts, $root;
  $arr = $points;
  $i = 0;
  while ($i < count($arr)) {
  $j = 0;
  while ($j < count($arr) - 1) {
  if ($arr[$j][$axis] > $arr[$j + 1][$axis]) {
  $tmp = $arr[$j];
  $arr[$j] = $arr[$j + 1];
  $arr[$j + 1] = $tmp;
}
  $j = $j + 1;
};
  $i = $i + 1;
};
  return $arr;
}
function build_kdtree($points, $depth) {
  global $tree, $pts, $root;
  if (count($points) == 0) {
  return 0 - 1;
}
  $k = count($points[0]);
  $axis = $depth % $k;
  $sorted = sort_points($points, $axis);
  $median_idx = count($sorted) / 2;
  $left_points = array_slice($sorted, 0, $median_idx - 0);
  $right_points = array_slice($sorted, $median_idx + 1, count($sorted) - ($median_idx + 1));
  $idx = count($tree);
  $tree = _append($tree, ['point' => $sorted[$median_idx], 'left' => 0 - 1, 'right' => 0 - 1]);
  $left_idx = build_kdtree($left_points, $depth + 1);
  $right_idx = build_kdtree($right_points, $depth + 1);
  $node = $tree[$idx];
  $node['left'] = $left_idx;
  $node['right'] = $right_idx;
  $tree[$idx] = $node;
  return $idx;
}
$pts = [[2.0, 3.0], [5.0, 4.0], [9.0, 6.0], [4.0, 7.0], [8.0, 1.0], [7.0, 2.0]];
$root = build_kdtree($pts, 0);
echo rtrim(_str($tree)), PHP_EOL;
echo rtrim(json_encode($root, 1344)), PHP_EOL;

File: tests/algorithms/transpiler/FS/data_structures/kd_tree/build_kdtree.fs
// Generated 2025-08-07 14:57 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
type KDNode = {
    point: float array
    left: int
    right: int
}
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let mutable tree: KDNode array = [||]
let rec sort_points (points: float array array) (axis: int) =
    let mutable __ret : float array array = Unchecked.defaultof<float array array>
    let mutable points = points
    let mutable axis = axis
    try
        let mutable arr: float array array = points
        let mutable i: int = 0
        while i < (Seq.length (arr)) do
            let mutable j: int = 0
            while j < ((Seq.length (arr)) - 1) do
                if (_idx (_idx arr (j)) (axis)) > (_idx (_idx arr (j + 1)) (axis)) then
                    let tmp: float array = _idx arr (j)
                    arr.[j] <- _idx arr (j + 1)
                    arr.[j + 1] <- tmp
                j <- j + 1
            i <- i + 1
        __ret <- arr
        raise Return
        __ret
    with
        | Return -> __ret
let rec build_kdtree (points: float array array) (depth: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable points = points
    let mutable depth = depth
    try
        if (Seq.length (points)) = 0 then
            __ret <- 0 - 1
            raise Return
        let k: int = Seq.length (_idx points (0))
        let axis: int = ((depth % k + k) % k)
        let sorted: float array array = sort_points (points) (axis)
        let median_idx: int = (Seq.length (sorted)) / 2
        let left_points: float array array = Array.sub sorted 0 (median_idx - 0)
        let right_points: float array array = Array.sub sorted (median_idx + 1) ((Seq.length (sorted)) - (median_idx + 1))
        let mutable idx: int = Seq.length (tree)
        tree <- Array.append tree [|{ point = _idx sorted (median_idx); left = 0 - 1; right = 0 - 1 }|]
        let left_idx: int = build_kdtree (left_points) (depth + 1)
        let right_idx: int = build_kdtree (right_points) (depth + 1)
        let mutable node: KDNode = _idx tree (idx)
        node <- { node with left = left_idx }
        node <- { node with right = right_idx }
        tree <- _arrset tree idx (node)
        __ret <- idx
        raise Return
        __ret
    with
        | Return -> __ret
let pts: float array array = [|[|2.0; 3.0|]; [|5.0; 4.0|]; [|9.0; 6.0|]; [|4.0; 7.0|]; [|8.0; 1.0|]; [|7.0; 2.0|]|]
let root: int = build_kdtree (pts) (0)
printfn "%s" (_str (tree))
printfn "%d" (root)
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
